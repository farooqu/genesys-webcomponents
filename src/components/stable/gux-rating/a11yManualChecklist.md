# gux-rating manual accessibility testing status
**Last Updated:** Wed Dec 22 2021 09:46:49 GMT-0500 (Eastern Standard Time)
| Pass | WCAG Success Criterion | Notes |
| --- | --- | --- |
| ❌ | [2.1.1 Keyboard](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html) | COMUI-822 A user would expect to be able to use the up and down arrows on an element with the role 'spinbutton' |
| ✅ | [2.1.2 No Keyboard Trap](https://www.w3.org/WAI/WCAG21/Understanding/no-keyboard-trap.html) | - |
| ✅ | [2.4.3 Focus Order](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html) | - |
| ❌ | [2.4.7 Focus Visible](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html) | COMUI-822 |
| ❌ | [2.5.3 Label in Name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html#dfn-name) | COMUI-822 The star rating value is not read to the screenreader |
| ✅ | [3.2.1 On Focus](https://www.w3.org/WAI/WCAG21/Understanding/on-focus.html) | - |
| ✅ | [3.2.2 On Input](https://www.w3.org/WAI/WCAG21/Understanding/on-input.html) | - |
| ✅ | [3.3.1 Error Identification](https://www.w3.org/WAI/WCAG21/Understanding/error-identification.html) | - |
| ✅ | [3.2.2 Labels or Instructions](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html) | - |
| ❌ | [4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html) | COMUI-822 Value of star rating should be announced to the screenreader each time it changes |
| ✅ | [4.1.3 Status Messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html) | - |