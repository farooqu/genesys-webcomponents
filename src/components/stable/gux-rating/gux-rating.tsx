import { Component, Element, h, Host, Listen, JSX, Prop } from '@stencil/core';

import simulateNativeEvent from '../../../utils/dom/simulate-native-event';
import clamp from '../../../utils/number/clamp';
import { trackComponent } from '../../../usage-tracking';

@Component({
  styleUrl: 'gux-rating.less',
  tag: 'gux-rating'
})
export class GuxRating {
  private ratingElement: HTMLDivElement;

  @Element()
  root: HTMLElement;

  @Prop({ mutable: true })
  value: number = 0;

  @Prop()
  maxValue: number = 5;

  @Prop()
  disabled: boolean = false;

  @Prop()
  readonly: boolean = false;

  @Listen('click')
  onClick(event: MouseEvent): void {
    event.stopPropagation();

    if (this.disabled || this.readonly) {
      return;
    }

    const ratingStar = (event.target as HTMLElement).closest('gux-icon');
    const clickedStarIndex = Array.from(this.ratingElement.children).findIndex(
      child => child === ratingStar
    );
    const clickedStarNominalValue = clickedStarIndex + 1;

    if (clickedStarNominalValue === this.value + 0.5) {
      this.updateRatingValue(clickedStarNominalValue);
    } else if (clickedStarNominalValue === this.value) {
      this.updateRatingValue(0);
    } else if (clickedStarNominalValue !== Math.floor(this.value)) {
      this.updateRatingValue(clickedStarNominalValue - 0.5);
    }
  }

  @Listen('keydown')
  onKeyDown(event: KeyboardEvent): void {
    event.stopPropagation();

    if (this.disabled || this.readonly) {
      return;
    }

    switch (event.key) {
      case 'ArrowUp':
      case 'ArrowLeft':
        event.preventDefault();
        this.updateRatingValue(this.value - 0.5);
        break;
      case 'ArrowDown':
      case 'ArrowRight':
        event.preventDefault();
        this.updateRatingValue(this.value + 0.5);
        break;
      case 'End':
        event.preventDefault();
        this.updateRatingValue(Infinity);
        break;
      case 'Home':
        event.preventDefault();
        this.updateRatingValue(-Infinity);
        break;
    }
  }

  private updateRatingValue(newValue: number): void {
    const validatedNewValue = clamp(
      newValue,
      0,
      Array.from(this.ratingElement.children).length
    );

    if (this.value !== validatedNewValue) {
      this.value = validatedNewValue;

      simulateNativeEvent(this.root, 'input');
      simulateNativeEvent(this.root, 'change');
    }
  }

  private getRatingStarElements(): JSX.Element {
    return [...Array(this.maxValue).keys()].reduce((acc, cv) => {
      if (cv + 0.5 === this.value) {
        return acc.concat(
          <gux-icon
            class="gux-rating-star-half"
            icon-name="rating-partial"
            decorative
          ></gux-icon>
        );
      } else if (cv + 1 <= this.value) {
        return acc.concat(
          <gux-icon
            class="gux-rating-star-full"
            icon-name="rating-active"
            decorative
          ></gux-icon>
        );
      }

      return acc.concat(
        <gux-icon
          class="gux-rating-star-empty"
          icon-name="rating"
          decorative
        ></gux-icon>
      );
    }, []);
  }

  private getTabIndex(): number {
    return this.disabled ? -1 : 0;
  }

  componentWillLoad(): void {
    trackComponent(this.root);
  }

  render(): JSX.Element {
    return (
      <Host
        role="spinbutton"
        tabindex={this.getTabIndex()}
        aria-readonly={this.readonly}
        aria-valuenow={this.value}
        aria-valuemin="0"
        aria-valuemax={this.maxValue}
      >
        <div
          ref={(el: HTMLDivElement) => (this.ratingElement = el)}
          class={{
            'gux-rating-start-container': true,
            'gux-disabled': this.disabled
          }}
        >
          {this.getRatingStarElements()}
        </div>
      </Host>
    );
  }
}
