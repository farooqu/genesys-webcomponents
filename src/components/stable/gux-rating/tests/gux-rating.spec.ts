jest.mock('../../../../utils/dom/random-html-id', () => ({
  randomHTMLId: () => 'random'
}));

import { newSpecPage } from '@stencil/core/testing';
import { GuxRating } from '../gux-rating';

global.InputEvent = Event;

const components = [GuxRating];
const language = 'en';

describe('gux-rating', () => {
  it('should build', async () => {
    const html = `<gux-rating aria-label="Feedback"></gux-rating>`;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxRating);
  });

  describe('#render', () => {
    [
      '<gux-rating aria-label="Feedback"></gux-rating>',
      '<gux-rating value="0" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="0.5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="1" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="1.5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="2" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="2.5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="3" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="3.5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="4" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="4.5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="5" aria-label="Feedback"></gux-rating>',
      '<gux-rating value="0" max-value=10 aria-label="Feedback"></gux-rating>',
      '<gux-rating value="5" max-value=10 aria-label="Feedback"></gux-rating>',
      '<gux-rating value="10" max-value=10 aria-label="Feedback"></gux-rating>',
      '<gux-rating value="3" disabled aria-label="Feedback"></gux-rating>',
      '<gux-rating value="3" readonly aria-label="Feedback"></gux-rating>'
    ].forEach((html, index) => {
      it(`should render component as expected (${index + 1})`, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });
  });
});
