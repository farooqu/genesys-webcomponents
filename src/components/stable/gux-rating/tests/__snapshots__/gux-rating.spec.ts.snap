// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-rating #render should render component as expected (1) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="0" role="spinbutton" tabindex="0">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (2) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="0" role="spinbutton" tabindex="0" value="0">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (3) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="0.5" role="spinbutton" tabindex="0" value="0.5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-half" decorative="" icon-name="rating-partial"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (4) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="1" role="spinbutton" tabindex="0" value="1">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (5) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="1.5" role="spinbutton" tabindex="0" value="1.5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-half" decorative="" icon-name="rating-partial"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (6) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="2" role="spinbutton" tabindex="0" value="2">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (7) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="2.5" role="spinbutton" tabindex="0" value="2.5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-half" decorative="" icon-name="rating-partial"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (8) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="3" role="spinbutton" tabindex="0" value="3">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (9) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="3.5" role="spinbutton" tabindex="0" value="3.5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-half" decorative="" icon-name="rating-partial"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (10) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="4" role="spinbutton" tabindex="0" value="4">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (11) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="4.5" role="spinbutton" tabindex="0" value="4.5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-half" decorative="" icon-name="rating-partial"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (12) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="5" role="spinbutton" tabindex="0" value="5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (13) 1`] = `
<gux-rating aria-valuemax="10" aria-valuemin="0" aria-valuenow="0" max-value="10" role="spinbutton" tabindex="0" value="0">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (14) 1`] = `
<gux-rating aria-valuemax="10" aria-valuemin="0" aria-valuenow="5" max-value="10" role="spinbutton" tabindex="0" value="5">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (15) 1`] = `
<gux-rating aria-valuemax="10" aria-valuemin="0" aria-valuenow="10" max-value="10" role="spinbutton" tabindex="0" value="10">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (16) 1`] = `
<gux-rating aria-valuemax="5" aria-valuemin="0" aria-valuenow="3" disabled="" role="spinbutton" tabindex="-1" value="3">
  <div class="gux-disabled gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;

exports[`gux-rating #render should render component as expected (17) 1`] = `
<gux-rating aria-readonly="" aria-valuemax="5" aria-valuemin="0" aria-valuenow="3" readonly="" role="spinbutton" tabindex="0" value="3">
  <div class="gux-rating-start-container">
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-full" decorative="" icon-name="rating-active"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
    <gux-icon class="gux-rating-star-empty" decorative="" icon-name="rating"></gux-icon>
  </div>
</gux-rating>
`;
